system/datatypes/mailinglist.php

```diff
@@ -0,0 +1,728 @@
+			$this->replacePageListNode(hypha_getPage($this->language, $this->pagename));
+			// check if pagename, privateFlag or language (in case of a new translation) have changed
+			if ($language != $this->language || $pagename != $this->pagename || $private != $this->privateFlag) {
+				hypha_setPage($this->pageListNode, $language, $pagename, $private);
+				$hyphaXml->saveAndUnlock();
+			} else {
+				$hyphaXml->unlock();
+			}
+			// unfortunately wymeditor can't handle relative urls so we'll add the baseUrl before editing and remove it afterwards
+			$this->xml->lockAndReload();
+			storeWikiContent($this->getDoc(), $language, wikify_html($form->dataFor(self::FIELD_NAME_DESCRIPTION)), $hyphaUser->getAttribute('username'));
+			$this->xml->saveAndUnlock();
+
+			notify('success', __('ml_successfully_updated'));
+			$this->redirect($pagename, $language);
+		}
+
+		// update the form dom so that error can be displayed, if there are any
+		$form->updateDom();
+
+		$this->findBySelector('#main')->append($form->elem->children());
+	}
+
+	private function confirmAction() {
+		$email = isset($_GET['email']) ? $_GET['email'] : null;
+		$code = isset($_GET['code']) ? $_GET['code'] : null;
+		if (!$email || !$code) {
+			return notify('error', __('missing-arguments'));
+		}
+
+		$address = $this->getDoc()->findXPath('//address[@email="' . $email . '"][@confirm-code="' . $code . '"]')->first();
+		if (!$address instanceof HyphaDomElement) {
+			return notify('error', __('invalid-email-code-combination'));
+		}
+
+		if ($address->getAttribute('status') == self::ADDRESS_STATUS_CONFIRMED) {
+			notify('success', __('ml_successfully_subscribed'));
+			$this->redirect('');
+		}
+
+		$id = $address->getId();
+		$this->xml->lockAndReload();
+		$address = $this->xml->document()->getElementById($id);
+		$address->setAttribute('status', self::ADDRESS_STATUS_CONFIRMED);
+		$address->setAttribute('unsubscribe-code', uniqid());
+		$this->getDoc()->getOrCreate('addresses')->append($address);
+		$this->xml->saveAndUnlock();
+
+		notify('success', __('ml_successfully_subscribed'));
+		$this->redirect('');
+	}
+
+	private function unsubscribeAction() {
+		$email = isset($_GET['email']) ? $_GET['email'] : null;
+		$code = isset($_GET['code']) ? $_GET['code'] : null;
+		if (!$email || !$code) {
+			return notify('error', __('missing-arguments'));
+		}
+
+		$address = $this->getDoc()->findXPath('//address[@email="' . $email . '"][@unsubscribe-code="' . $code . '"]')->first();
+		if (!$address instanceof HyphaDomElement) {
+			return notify('error', __('invalid-email-code-combination'));
+		}
+
+		if ($address->getAttribute('status') != self::ADDRESS_STATUS_CONFIRMED) {
+			return notify('error', __('email-has-invalid-status'));
+			$this->redirect('');
+		}
+
+		$id = $address->getId();
+		$this->xml->lockAndReload();
+		$address = $this->xml->document()->getElementById($id);
+		$this->getDoc()->getOrCreate('addresses')->removeChild($address);
```

Comment:
Although I'm not a big fan of communicating id's, I do think this is a good solution. But I would rather communicate 'unsubscribed' instead of 'already unsubscribed'.

Reply:
As before, I do not have a strong preference over 'unsubscribed' vs 'already unsubscribed'.